File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.java
class Solution {
    public int characterReplacement(String s, int k) {
        
        HashMap<Character, Integer> map = new HashMap<>();

        int left = 0;
        int result = 0;
        int maxFreq = 0; // To track the frequency of the most frequent character in the current window

        for (int right = 0; right < s.length(); right++) {
            char eachChar = s.charAt(right);

            // Add the current character to the hashmap
            map.merge(eachChar, 1, Integer::sum);

            // Update maxFreq to reflect the frequency of the most frequent character in the window
            maxFreq = Math.max(maxFreq, map.get(eachChar));

            // If the current window size minus the max frequency is greater than k, shrink the window
            while ((right - left + 1) - maxFreq > k) {
                char leftChar = s.charAt(left);
                map.put(leftChar, map.get(leftChar) - 1);
                left++;
            }

            // Update the result with the size of the current valid window
            result = Math.max(result, right - left + 1);
        }

        return result;

    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right, result, seen = 0, 0, 0, set()

        while right < len(s):
            right_char = s[right]

            while right_char in seen and left <= right:
                seen.remove(s[left])
                left += 1

            seen.add(right_char)
            right += 1
            result = max(result, right - left)

        return max(result, right - left)